src/bot.rs

```diff
@@ -0,0 +1,44 @@
+use engine::GameState;
+
+pub trait Bot {
+    fn tick(&mut self, game_state: &GameState) -> ActionSet;
+    fn game_start(&mut self) {}
+    fn game_end(&mut self) {}
+}
+
+impl<F> Bot for F
+where
+    F: FnMut(&GameState) -> ActionSet,
+{
+    fn tick(&mut self, game_state: &GameState) -> ActionSet {
+        self(game_state)
+    }
+}
+
+
+/// A set of queued actions to be executed on the next tick.
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct ActionSet {
+    actions: Vec<Action>,
+}
+
+impl ActionSet {
+    pub fn empty() -> ActionSet {
+        ActionSet::default()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.actions.is_empty()
+    }
+
+    pub(crate) fn actions(&self) -> ::std::slice::Iter<Action> {
```

Comment:
I've not seen `pub(crate)` before what does this do?

Reply:
It means that this thing is `pub`, but only crate-wide. So then *we* can access the `actions()` easily, but it's still hidden from general users.

[relevant rfc](https://github.com/nox/rust-rfcs/blob/master/text/1422-pub-restricted.md)